srtcore/api.cpp

```diff
@@ -2666,6 +2723,13 @@ void srt::CUDTUnited::checkBrokenSockets()
     for (sockets_t::iterator j = m_ClosedSockets.begin(); j != m_ClosedSockets.end(); ++j)
     {
         CUDTSocket* ps = j->second;
+
+        if (ps->isStillBusy())
+        {
+            HLOGC(smlog.Debug, log << "checkBrokenSockets: @" << ps->m_SocketID << " is still busy, SKIPPING THIS CYCLE.");
+            continue;
+        }
+
```

Comment:
There is still a hypothetical issue here (even if the code does not do it as of now) that after you check if the socket is busy or not, you don't block possible acquisition later on. Ideally it should be `acquireIfNotBusy()` uniquely blocking the socket for destruction.

Reply:
Well, but I don't think such a scenario is at all possible because:
1. The socket gets moved to m_ClosedSocket, and it still waits for the GC to pick it up.
2. Such a socket is no longer dispatchable, so since this moment either some other facility has already acquired it BEFORE IT HAPPENED, or it will no longer have a chance.
3. This means that if the socket ALREADY IS in `m_ClosedSockets`, no facility can acquire this socket anymore.

This is also guaranteed by the fact that moving to `m_ClosedSockets` can only happen under guard of `m_GlobControlLock` and this whole code is also under this lock.

There is a risk when socket acquisition bypasses dispatching, when the constructor of SocketKeeper takes it by pointer, but this then executes `CUDTUnited::acquireSocket`, which disallows acquisition of a closed socket. Note that the sole use of pointers to sockets is also dangerous by definition and therefore all facilities that keep sockets by pointers must  provide a reverse information about the presence of the socket in these facilities, be it even through the acquisition flag.

Reply:
Ok, for that matter, probably the SocketKeeper should be embedded in all facilities that keep sockets by pointer, such as CSndUList, for example.

Reply:
```suggestion
        // There is still a hypothetical risk here (even if the code does not do it as of now)
        // that after ps is checked for being busy or not, its possible acquisition later on is not blocked.
        // Ideally it should be acquireIfNotBusy() uniquely blocking the socket for destruction.
        if (ps->isStillBusy())
        {
            HLOGC(smlog.Debug, log << "checkBrokenSockets: @" << ps->m_SocketID << " is still busy, SKIPPING THIS CYCLE.");
            continue;
        }

```